### PR TITLE
*: forbid set TiFlash Replica for a table with unsupport charset

### LINKED ISSUE #30095: Forbid set TiFlashReplica for a table with GBK charset
Currently, TiFlash does not support GBK-related features. Therefore, TiFlashReplica should not be set for the table property with GBK.

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #30095

Problem Summary: forbid set TiFlash Replica for a table with unsupport charset

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Defined2014** [comment]
/cc @tangenta @xiongjiwei @zimulala @AilinKid @wjhuang2016 
PTAL

File: parser/charset/charset.go [hunk 1]
```diff
@@ -68,6 +68,15 @@ var supportedCollationNames = map[string]struct{}{
 	CollationBin:     {},
 }
 
+// TiFlashSupportedCharsets is a map which contains TiFlash supports charsets.
+var TiFlashSupportedCharsets = map[string]bool{
```
Comment (tangenta) [inline parser/charset/charset.go:72]: I think a `map[string]struct{}` is enough?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -4721,6 +4721,14 @@ func (d *ddl) AlterTableSetTiFlashReplica(ctx sessionctx.Context, ident ast.Iden
 		return ErrOptOnTemporaryTable.GenWithStackByArgs("set tiflash replica")
 	}
 
+	// Ban setting replica count for tables which has charset not supported by TiFlash
+	for _, col := range tb.Cols() {
+		_, v := charset.TiFlashSupportedCharsets[col.Charset]
+		if !v {
```
Comment (tangenta) [inline ddl/ddl_api.go:4727]:
```suggestion
		_, ok := charset.TiFlashSupportedCharsets[col.Charset]
		if !ok {
```

Comment (Defined2014) [reply ddl/ddl_api.go]: Done. @tangenta

File: executor/tiflash_test.go [hunk 3]
```diff
@@ -24,6 +24,8 @@ import (
 	"sync/atomic"
 	"time"
 
+	"github.com/pingcap/tidb/util/collate"
+
```
Comment (tangenta) [inline executor/tiflash_test.go:28]: Please format these import directives :)

Comment (Defined2014) [reply executor/tiflash_test.go]: Done. @tangenta

**Review by tangenta — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**